### PR TITLE
Rework on the example, include cuFFT and bug correction for CUDA-aware MPI/nccl

### PR DESCRIPTION
**Pull request by rfj82982**
Hi Guys I have put all the work on cuFFT and CUDA-aware MPI together and I have also merged with the latest version of the main. A substantial change is the rework of the examples now only 4 are available: halo, 2dtest, fft_physical in Z and fft_physical in Z. All of them works also on GPU however only 3 (the 2 fft and the transpose) have been really modified to work on GPU. 
i tested them using gnu 12 with mpich(mac) and openmpi with both generic and FFTW. For the GPU i tested NVHPC 22.7 with also NCCL. I would suggest for you to have a good look since changes might me quite large.

### CODE REVIEW

**Comment by mathrack** [comment]
Thank you ! I will try to look at it deeply shortly. Please update

- [ ] the Makefile, I think it does not cover the cufft back-end yet
- [ ] the README to describe the GPU / NCCL preprocessor variables and the new optional dependency (cuFFT)
- [ ] please add error checks on each call to cuda functions

Cheers

**Comment by rfj82982** [comment]
The back-end for the cuFFT works. I have a direct path in the Makefile that i have to make more general. The remaining is with the compiler flags as well as the NCCL. For all these libs only the INC is needed

File: examples/fft_physical_x/Makefile [hunk 1]
```diff
@@ -5,19 +5,27 @@ FFLAGS := $(subst $(MODFLAG),$(MODFLAG)../../,$(FFLAGS))
...
+	$(MPIRUN) -n $(NP) ./bind.sh ./fft_physical_x
+else
 check:
-	mpirun -n 4 ./timing
+	$(MPIRUN) -n $(NP) ./fft_physical_x
+endif
 
 mem_leak:
-	valgrind --leak-check=full --show-leak-kinds=all mpirun -n 1 ./timing 1 1
+	valgrind --leak-check=full --show-leak-kinds=all mpirun -n 1 ./fft_physical_x 1 1
```
Comment (mathrack) [inline examples/fft_physical_x/Makefile:25]: please replace `mpirun` with $(MPIRUN)

Comment (mathrack) [reply examples/fft_physical_x/Makefile:25]: Similar comment for fft_physical_in_z

**Comment by mathrack** [comment]
Using the nvidia compiler, is it possible to automatic compile the code with cufft and the GPU / NCCL flags ?

**Comment by rfj82982** [comment]
Yes I am writing how to do this in the README. Just use the following
`make CMP=nvhpc FFT=cufft PARAMOD=gpu CUFFT_PATH=/work4/scd/scarf562/eb-amd/software/NVHPC/22.7/Linux_x86_64/22.7/compilers`

**Comment by rfj82982** [comment]
CUDA-aware mpi and NCCL are activated by the FLAGS `_GPU` and `_NCCL`

**Comment by mathrack** [comment]
I have tried to build the code with the provided `make` command but it keeps failing. See [this](https://github.com/xcompact3d/2decomp-fft/actions/runs/3136181962/jobs/5092773158) for instance.

I will try to fix it shortly, or will remove the GPU build from the automatic tests

**Comment by mathrack** [comment]
A few extra minor comments.

I am not sure removing the IO tests is sharp because they are fixed and used in https://github.com/xcompact3d/2decomp-fft/pull/20. And the IO branch, when ready, will be merged. However, the tests are present there. Thus, even if we remove them here, we can restore them later. To be discussed.

In the test `fft_physical_in_x`, the following calling sequence is performed : `decomp_2d_fft_init`, test complex FFT, `decomp_2d_fft_finalize`, `decomp_2d_fft_init`, test real-complex and complex-real FFT, `decomp_2d_fft_finalize`. In my opinion, the expected usage for the 2decomp is `decomp_2d_fft_init` at the beginning of the program and `decomp_2d_fft_finalize` at the end. I am not sure it is sharp to test a program performing several fft_init / fft_finalize.

File: examples/fft_physical_x/Makefile [hunk 2]
```diff
@@ -5,22 +5,27 @@ FFLAGS := $(subst $(MODFLAG),$(MODFLAG)../../,$(FFLAGS))
 FFLAGS := $(patsubst -I%,-I../../%,$(FFLAGS))
 LIBS = -L../../ -l$(LIBDECOMP) $(LIBFFT) $(LFLAGS)
 
-OBJ = fft_test_r2c.o
+OBJ = fft_physical_x.o
 
-NP ?= 1
+NP ?= 4
```
Comment (pbartholomew08) [inline examples/fft_physical_x/Makefile:10]: Default to 1

File: examples/fft_physical_z/Makefile [hunk 3]
```diff
@@ -5,25 +5,30 @@ FFLAGS := $(subst $(MODFLAG),$(MODFLAG)../../,$(FFLAGS))
 FFLAGS := $(patsubst -I%,-I../../%,$(FFLAGS))
 LIBS = -L../../ -l$(LIBDECOMP) $(LIBFFT) $(LFLAGS)
 
-OBJ = timing.o
+OBJ = fft_physical_z.o
 
-timing: $(OBJ)
+NP ?= 4
```
Comment (pbartholomew08) [inline examples/fft_physical_z/Makefile:10]: Defauly to 1

File: examples/test2d/Makefile [hunk 4]
```diff
@@ -8,14 +8,20 @@ LIBS = -L../../ -l$(LIBDECOMP) $(LFLAGS)
 
 OBJ = test2d.o
 
-NP ?= 1
+NP ?= 2
```
Comment (pbartholomew08) [inline examples/test2d/Makefile:11]: Default to 1

File: src/transpose_z_to_y.f90 [hunk 5]
```diff
@@ -25,6 +25,7 @@ subroutine transpose_z_to_y_real(src, dst, opt_decomp)
 #if defined(_NCCL)
     integer :: row_rank_id
 #endif
+    integer :: istat
```
Comment (pbartholomew08) [inline src/transpose_z_to_y.f90:28]: Guard definition by ```_GPU```

**Review by pbartholomew08 — CHANGES_REQUESTED** [review]

File: src/transpose_z_to_y.f90 (hunk 5 above)
Comment (pbartholomew08) [reply src/transpose_z_to_y.f90:28]: My mistake - it is guarded

**Review by pbartholomew08 — APPROVED** [review]

**Pull request merged**